docs/design/metainfo-rpc.md

```diff
@@ -0,0 +1,129 @@
+# Metainfo RPC Service Refactor
+
+## Background
+
+Our current metainfo rpc service isn't very future proof.
+If we shipped our current client code, it wouldn't be easy to update clients and add support for the following:
+- versions
+- multipart uploads
+- cleanup for uncommitted segments
+- general concurrency fixes
+
+Multipart upload can't happen with our current architecture because there's no concept of starting and ending a stream.
+
+Versions also wouldn't work, because if concurrent clients tried to upload to the same version, they would clobber each other.
```

Comment:
Maybe s/to the same version/at the same time/. The problem is that the current protobuf does not contain which version is being written, so it has to assume "latest version", which is problematic with concurrent clients (they both are writing to the "latest version").

Reply:
I just added what you said here